pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,2003 @@
+    MPIManager mpi_manager_;
+
+    SharedCusvHandle handle_;
+    SharedCublasCaller cublascaller_;
+    mutable SharedCusparseHandle
+        cusparsehandle_; // This member is mutable to allow lazy initialization.
+    SharedLocalStream localStream_;
+    SharedMPIWorker svSegSwapWorker_;
+    GateCache<Precision> gate_cache_;
+
+  public:
+    using CFP_t =
+        typename StateVectorCudaBase<Precision,
+                                     StateVectorCudaMPI<Precision>>::CFP_t;
+    using GateType = CFP_t *;
+
+    StateVectorCudaMPI() = delete;
+
+    StateVectorCudaMPI(MPIManager mpi_manager, size_t num_global_qubits,
+                       size_t num_local_qubits)
+        : StateVectorCudaBase<Precision, StateVectorCudaMPI<Precision>>(
+              num_local_qubits),
+          numGlobalQubits_(num_global_qubits),
+          numLocalQubits_(num_local_qubits), mpi_manager_(mpi_manager),
+          handle_(make_shared_cusv_handle()),
+          cublascaller_(make_shared_cublas_caller()),
+          localStream_(make_shared_local_stream()),
+          svSegSwapWorker_(make_shared_mpi_worker<CFP_t>(
+              handle_.get(), mpi_manager_, BaseType::getData(),
+              num_local_qubits, localStream_.get())),
+          gate_cache_(true) {
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize())
+        mpi_manager_.Barrier();
+    };
+
+    StateVectorCudaMPI(MPI_Comm mpi_communicator, size_t num_global_qubits,
+                       size_t num_local_qubits)
+        : StateVectorCudaBase<Precision, StateVectorCudaMPI<Precision>>(
+              num_local_qubits),
+          numGlobalQubits_(num_global_qubits),
+          numLocalQubits_(num_local_qubits), mpi_manager_(mpi_communicator),
+          handle_(make_shared_cusv_handle()),
+          cublascaller_(make_shared_cublas_caller()),
+          localStream_(make_shared_local_stream()),
+          svSegSwapWorker_(make_shared_mpi_worker<CFP_t>(
+              handle_.get(), mpi_manager_, BaseType::getData(),
+              num_local_qubits, localStream_.get())),
+          gate_cache_(true) {
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize())
+        mpi_manager_.Barrier();
+    };
+
+    StateVectorCudaMPI(size_t num_global_qubits, size_t num_local_qubits)
+        : StateVectorCudaBase<Precision, StateVectorCudaMPI<Precision>>(
+              num_local_qubits),
+          numGlobalQubits_(num_global_qubits),
+          numLocalQubits_(num_local_qubits), mpi_manager_(MPI_COMM_WORLD),
+          handle_(make_shared_cusv_handle()),
+          cublascaller_(make_shared_cublas_caller()),
+          localStream_(make_shared_local_stream()),
+          svSegSwapWorker_(make_shared_mpi_worker<CFP_t>(
+              handle_.get(), mpi_manager_, BaseType::getData(),
+              num_local_qubits, localStream_.get())),
+          gate_cache_(true) {
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize())
+        mpi_manager_.Barrier();
+    };
+
+    ~StateVectorCudaMPI(){};
```

Comment:
I guess a few things are still on `origin/mpi_gate_operation`, e.g.
```suggestion
    ~StateVectorCudaMPI() = default;
```